Misinterpreted closure parameter when wrapped in parentheses

## New Issue Checklist

- [x] I updated my Mockingbird framework and CLI to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/birdrides/mockingbird/issues)

## Description

Extra parens around closure parameter breaks mocked method.

### Generator Bugs

```bash
$ mockingbird generate \
  --only-protocols \
  --targets 'Sample' \
  --outputs 'Sample/MockingbirdMocks/SampleMocks.generated.swift'
```

1. A minimal example of the original source

```swift
protocol One : class {
    func foo(_ x: ((Int) -> Void))
}
```

2. The actual mocking code generated

```swift
//
//  SampleMocks.generated.swift
//  Sample
//
//  Generated by Mockingbird v0.9.0.
//  DO NOT EDIT
//

@testable import Mockingbird
@testable import Sample
import Foundation
import Swift
import UIKit

private var genericTypesStaticMocks = Mockingbird.Synchronized<[String: Mockingbird.StaticMock]>([:])

// MARK: - Mocked One

public final class OneMock: Sample.One, Mockingbird.Mock {
  static let staticMock = Mockingbird.StaticMock()
  public let mockingContext = Mockingbird.MockingContext()
  public let stubbingContext = Mockingbird.StubbingContext()
  public let mockMetadata = Mockingbird.MockMetadata(["generator_version": "0.9.0", "module_name": "Sample"])
  public var sourceLocation: Mockingbird.SourceLocation? {
    get { return stubbingContext.sourceLocation }
    set {
      stubbingContext.sourceLocation = newValue
      OneMock.staticMock.stubbingContext.sourceLocation = newValue
    }
  }

  public enum InitializerProxy {}

  // MARK: Mocked `foo`(_ x: ((Int) -> Void))

  public func `foo`(_ x: ((Int) -> Void)) -> Void {
    let invocation: Mockingbird.Invocation = Mockingbird.Invocation(selectorName: "`foo`(_ x: ((Int) -> Void)) -> Void", arguments: [Mockingbird.ArgumentMatcher(`x`)])
    mockingContext.didInvoke(invocation)
    let implementation = stubbingContext.implementation(for: invocation, optional: true)
    if let concreteImplementation = implementation as? (((Int) -> Void)) -> Void {
      concreteImplementation(`x`)
    } else {
      (implementation as? () -> Void)?()
    }
  }

  public func `foo`(_ x: @escaping @autoclosure () -> ((Int) -> Void)) -> Mockingbird.Mockable<Mockingbird.MethodDeclaration, (((Int) -> Void)) -> Void, Void> {
    let arguments: [Mockingbird.ArgumentMatcher] = [Mockingbird.resolve(`x`)]
    let invocation: Mockingbird.Invocation = Mockingbird.Invocation(selectorName: "`foo`(_ x: ((Int) -> Void)) -> Void", arguments: arguments)
    return Mockingbird.Mockable<Mockingbird.MethodDeclaration, (((Int) -> Void)) -> Void, Void>(mock: self, invocation: invocation)
  }
}

/// Create a source-attributed `Sample.One` class mock metatype.
public func mock(file: StaticString = #file, line: UInt = #line, _ type: Sample.One.Protocol) -> OneMock.InitializerProxy.Type {
  return OneMock.InitializerProxy.self
}
```

Compilation error:
```
SampleMocks.generated.swift:37:134: error: converting non-escaping value to 'T' may allow it to escape
    let invocation: Mockingbird.Invocation = Mockingbird.Invocation(selectorName: "`foo`(_ x: ((Int) -> Void)) -> Void", arguments: [Mockingbird.ArgumentMatcher(`x`)])
```

3. The expected mocking code that should be generated (or a description)

```swift
  // MARK: Mocked `foo`(_ x: (Int) -> Void)

  public func `foo`(_ x: (Int) -> Void) -> Void {
    let invocation: Mockingbird.Invocation = Mockingbird.Invocation(selectorName: "`foo`(_ x: (Int) -> Void) -> Void", arguments: [Mockingbird.ArgumentMatcher(Mockingbird.NonEscapingClosure<(Int) -> Void>())])
```

## Environment

* Mockingbird CLI version `0.9.0`
* Xcode `11.3.1` and macOS version `10.15.3`
* Swift version `Apple Swift version 5.1.3 (swiftlang-1100.0.282.1 clang-1100.0.33.15)`
* Installation method `CocoaPods`

--- a/MockingbirdGenerator/Parser/Models/MethodParameter.swift
+++ b/MockingbirdGenerator/Parser/Models/MethodParameter.swift
@@ -51,9 +51,26 @@ struct MethodParameter: Hashable {
     
     // Final attributes can differ from those in `parameter` due to knowing the typealiased type.
     var attributes = Attributes(from: dictionary).union(actualParameter.attributes)
-    if actualParameter.type.isFunction { attributes.insert(.closure) }
+    if actualParameter.type.isFunction || actualParameter.type.isWrappedClosure {
+      attributes.insert(.closure)
+    }
     self.typeName = typeName
     self.attributes = attributes
     self.hasSelfConstraints = typeName.contains(SerializationRequest.Constants.selfTokenIndicator)
+  }
+}
+
+private extension DeclaredType {
+  /// Whether the type is a non-escaping closure wrapped in a single (non-optional) element tuple.
+  var isWrappedClosure: Bool {
+    switch self {
+    case .single: return false
+    case .tuple(let tuple, _):
+      guard
+        !isOptional, // Closures in optionals are implicitly escaping.
+        tuple.elements.count == 1 // Closures in multiple element tuples are implicitly escaping.
+        else { return false }
+      return tuple.elements.first?.type.isFunction == true
+    }
   }
 }

--- a/MockingbirdMocks/MockingbirdTestsHostMocks.generated.swift
+++ b/MockingbirdMocks/MockingbirdTestsHostMocks.generated.swift
@@ -3045,6 +3045,44 @@ public final class ClosureParametersProtocolMock: MockingbirdTestsHost.ClosurePa
     return Mockingbird.Mockable<Mockingbird.MethodDeclaration, (@escaping () -> Bool) -> Bool, Bool>(mock: self, invocation: invocation)
   }
 
+  // MARK: Mocked `implicitEscapingMultipleLabeledTupleClosure`(block: (a: () -> Void, b: (Bool) -> Bool))
+
+  public func `implicitEscapingMultipleLabeledTupleClosure`(block: (a: () -> Void, b: (Bool) -> Bool)) -> Bool {
+    let invocation: Mockingbird.Invocation = Mockingbird.Invocation(selectorName: "`implicitEscapingMultipleLabeledTupleClosure`(block: (a: () -> Void, b: (Bool) -> Bool)) -> Bool", arguments: [Mockingbird.ArgumentMatcher(`block`)])
+    mockingContext.didInvoke(invocation)
+    let implementation = stubbingContext.implementation(for: invocation, optional: false)
+    if let concreteImplementation = implementation as? ((a: () -> Void, b: (Bool) -> Bool)) -> Bool {
+      return concreteImplementation(`block`)
+    } else {
+      return (implementation as! () -> Bool)()
+    }
+  }
+
+  public func `implicitEscapingMultipleLabeledTupleClosure`(block: @escaping @autoclosure () -> (a: () -> Void, b: (Bool) -> Bool)) -> Mockingbird.Mockable<Mockingbird.MethodDeclaration, ((a: () -> Void, b: (Bool) -> Bool)) -> Bool, Bool> {
+    let arguments: [Mockingbird.ArgumentMatcher] = [Mockingbird.resolve(`block`)]
+    let invocation: Mockingbird.Invocation = Mockingbird.Invocation(selectorName: "`implicitEscapingMultipleLabeledTupleClosure`(block: (a: () -> Void, b: (Bool) -> Bool)) -> Bool", arguments: arguments)
+    return Mockingbird.Mockable<Mockingbird.MethodDeclaration, ((a: () -> Void, b: (Bool) -> Bool)) -> Bool, Bool>(mock: self, invocation: invocation)
+  }
+
+  // MARK: Mocked `implicitEscapingMultipleTupleClosure`(block: (() -> Void, (Bool) -> Bool))
+
+  public func `implicitEscapingMultipleTupleClosure`(block: (() -> Void, (Bool) -> Bool)) -> Bool {
+    let invocation: Mockingbird.Invocation = Mockingbird.Invocation(selectorName: "`implicitEscapingMultipleTupleClosure`(block: (() -> Void, (Bool) -> Bool)) -> Bool", arguments: [Mockingbird.ArgumentMatcher(`block`)])
+    mockingContext.didInvoke(invocation)
+    let implementation = stubbingContext.implementation(for: invocation, optional: false)
+    if let concreteImplementation = implementation as? ((() -> Void, (Bool) -> Bool)) -> Bool {
+      return concreteImplementation(`block`)
+    } else {
+      return (implementation as! () -> Bool)()
+    }
+  }
+
+  public func `implicitEscapingMultipleTupleClosure`(block: @escaping @autoclosure () -> (() -> Void, (Bool) -> Bool)) -> Mockingbird.Mockable<Mockingbird.MethodDeclaration, ((() -> Void, (Bool) -> Bool)) -> Bool, Bool> {
+    let arguments: [Mockingbird.ArgumentMatcher] = [Mockingbird.resolve(`block`)]
+    let invocation: Mockingbird.Invocation = Mockingbird.Invocation(selectorName: "`implicitEscapingMultipleTupleClosure`(block: (() -> Void, (Bool) -> Bool)) -> Bool", arguments: arguments)
+    return Mockingbird.Mockable<Mockingbird.MethodDeclaration, ((() -> Void, (Bool) -> Bool)) -> Bool, Bool>(mock: self, invocation: invocation)
+  }
+
   // MARK: Mocked `implicitEscapingParameterizedClosure`(block: ((Bool) -> Void)?)
 
   public func `implicitEscapingParameterizedClosure`(block: ((Bool) -> Void)?) -> Bool {
@@ -3159,6 +3197,44 @@ public final class ClosureParametersProtocolMock: MockingbirdTestsHost.ClosurePa
     return Mockingbird.Mockable<Mockingbird.MethodDeclaration, ((Bool) -> Bool) -> Bool, Bool>(mock: self, invocation: invocation)
   }
 
+  // MARK: Mocked `parameterizedReturningTupleClosure`(block: ((Bool) -> Bool))
+
+  public func `parameterizedReturningTupleClosure`(block: ((Bool) -> Bool)) -> Bool {
+    let invocation: Mockingbird.Invocation = Mockingbird.Invocation(selectorName: "`parameterizedReturningTupleClosure`(block: ((Bool) -> Bool)) -> Bool", arguments: [Mockingbird.ArgumentMatcher(Mockingbird.NonEscapingClosure<((Bool) -> Bool)>())])
+    mockingContext.didInvoke(invocation)
+    let implementation = stubbingContext.implementation(for: invocation, optional: false)
+    if let concreteImplementation = implementation as? (((Bool) -> Bool)) -> Bool {
+      return concreteImplementation(`block`)
+    } else {
+      return (implementation as! () -> Bool)()
+    }
+  }
+
+  public func `parameterizedReturningTupleClosure`(block: @escaping @autoclosure () -> ((Bool) -> Bool)) -> Mockingbird.Mockable<Mockingbird.MethodDeclaration, (((Bool) -> Bool)) -> Bool, Bool> {
+    let arguments: [Mockingbird.ArgumentMatcher] = [Mockingbird.resolve(`block`)]
+    let invocation: Mockingbird.Invocation = Mockingbird.Invocation(selectorName: "`parameterizedReturningTupleClosure`(block: ((Bool) -> Bool)) -> Bool", arguments: arguments)
+    return Mockingbird.Mockable<Mockingbird.MethodDeclaration, (((Bool) -> Bool)) -> Bool, Bool>(mock: self, invocation: invocation)
+  }
+
+  // MARK: Mocked `parameterizedTupleClosure`(block: ((Bool) -> Void))
+
+  public func `parameterizedTupleClosure`(block: ((Bool) -> Void)) -> Bool {
+    let invocation: Mockingbird.Invocation = Mockingbird.Invocation(selectorName: "`parameterizedTupleClosure`(block: ((Bool) -> Void)) -> Bool", arguments: [Mockingbird.ArgumentMatcher(Mockingbird.NonEscapingClosure<((Bool) -> Void)>())])
+    mockingContext.didInvoke(invocation)
+    let implementation = stubbingContext.implementation(for: invocation, optional: false)
+    if let concreteImplementation = implementation as? (((Bool) -> Void)) -> Bool {
+      return concreteImplementation(`block`)
+    } else {
+      return (implementation as! () -> Bool)()
+    }
+  }
+
+  public func `parameterizedTupleClosure`(block: @escaping @autoclosure () -> ((Bool) -> Void)) -> Mockingbird.Mockable<Mockingbird.MethodDeclaration, (((Bool) -> Void)) -> Bool, Bool> {
+    let arguments: [Mockingbird.ArgumentMatcher] = [Mockingbird.resolve(`block`)]
+    let invocation: Mockingbird.Invocation = Mockingbird.Invocation(selectorName: "`parameterizedTupleClosure`(block: ((Bool) -> Void)) -> Bool", arguments: arguments)
+    return Mockingbird.Mockable<Mockingbird.MethodDeclaration, (((Bool) -> Void)) -> Bool, Bool>(mock: self, invocation: invocation)
+  }
+
   // MARK: Mocked `trivialClosure`(block: () -> Void)
 
   public func `trivialClosure`(block: () -> Void) -> Bool {
@@ -3195,6 +3271,44 @@ public final class ClosureParametersProtocolMock: MockingbirdTestsHost.ClosurePa
     let arguments: [Mockingbird.ArgumentMatcher] = [Mockingbird.resolve(`block`)]
     let invocation: Mockingbird.Invocation = Mockingbird.Invocation(selectorName: "`trivialReturningClosure`(block: () -> Bool) -> Bool", arguments: arguments)
     return Mockingbird.Mockable<Mockingbird.MethodDeclaration, (() -> Bool) -> Bool, Bool>(mock: self, invocation: invocation)
+  }
+
+  // MARK: Mocked `trivialReturningTupleClosure`(block: (() -> Void))
+
+  public func `trivialReturningTupleClosure`(block: (() -> Void)) -> Bool {
+    let invocation: Mockingbird.Invocation = Mockingbird.Invocation(selectorName: "`trivialReturningTupleClosure`(block: (() -> Void)) -> Bool", arguments: [Mockingbird.ArgumentMatcher(Mockingbird.NonEscapingClosure<(() -> Void)>())])
+    mockingContext.didInvoke(invocation)
+    let implementation = stubbingContext.implementation(for: invocation, optional: false)
+    if let concreteImplementation = implementation as? ((() -> Void)) -> Bool {
+      return concreteImplementation(`block`)
+    } else {
+      return (implementation as! () -> Bool)()
+    }
+  }
+
+  public func `trivialReturningTupleClosure`(block: @escaping @autoclosure () -> (() -> Void)) -> Mockingbird.Mockable<Mockingbird.MethodDeclaration, ((() -> Void)) -> Bool, Bool> {
+    let arguments: [Mockingbird.ArgumentMatcher] = [Mockingbird.resolve(`block`)]
+    let invocation: Mockingbird.Invocation = Mockingbird.Invocation(selectorName: "`trivialReturningTupleClosure`(block: (() -> Void)) -> Bool", arguments: arguments)
+    return Mockingbird.Mockable<Mockingbird.MethodDeclaration, ((() -> Void)) -> Bool, Bool>(mock: self, invocation: invocation)
+  }
+
+  // MARK: Mocked `trivialTupleClosure`(block: (() -> Void))
+
+  public func `trivialTupleClosure`(block: (() -> Void)) -> Bool {
+    let invocation: Mockingbird.Invocation = Mockingbird.Invocation(selectorName: "`trivialTupleClosure`(block: (() -> Void)) -> Bool", arguments: [Mockingbird.ArgumentMatcher(Mockingbird.NonEscapingClosure<(() -> Void)>())])
+    mockingContext.didInvoke(invocation)
+    let implementation = stubbingContext.implementation(for: invocation, optional: false)
+    if let concreteImplementation = implementation as? ((() -> Void)) -> Bool {
+      return concreteImplementation(`block`)
+    } else {
+      return (implementation as! () -> Bool)()
+    }
+  }
+
+  public func `trivialTupleClosure`(block: @escaping @autoclosure () -> (() -> Void)) -> Mockingbird.Mockable<Mockingbird.MethodDeclaration, ((() -> Void)) -> Bool, Bool> {
+    let arguments: [Mockingbird.ArgumentMatcher] = [Mockingbird.resolve(`block`)]
+    let invocation: Mockingbird.Invocation = Mockingbird.Invocation(selectorName: "`trivialTupleClosure`(block: (() -> Void)) -> Bool", arguments: arguments)
+    return Mockingbird.Mockable<Mockingbird.MethodDeclaration, ((() -> Void)) -> Bool, Bool>(mock: self, invocation: invocation)
   }
 
   // MARK: Mocked `wrappedClosureParameter`(block: MockingbirdTestsHost.ClosureWrapper)

--- a/MockingbirdTests/Framework/ClosureParameterTests.swift
+++ b/MockingbirdTests/Framework/ClosureParameterTests.swift
@@ -117,6 +117,52 @@ class ClosureParameterTests: XCTestCase {
     verify(concreteMock.wrappedClosureParameter(block: any())).wasCalled()
   }
   
+  // MARK: Tuple wrapped non-escaping
+  
+  func testTrivialTupleClosure_anyWildcardMatching() {
+    given(concreteMock.trivialTupleClosure(block: any())) ~> true
+    XCTAssertTrue((concreteMock as ClosureParametersProtocol)
+      .trivialTupleClosure(block: {}))
+    verify(concreteMock.trivialTupleClosure(block: any())).wasCalled()
+  }
+  
+  func testTrivialReturningTupleClosure_anyWildcardMatching() {
+    given(concreteMock.trivialReturningTupleClosure(block: any())) ~> true
+    XCTAssertTrue((concreteMock as ClosureParametersProtocol)
+      .trivialReturningTupleClosure(block: { fatalError() }))
+    verify(concreteMock.trivialReturningTupleClosure(block: any())).wasCalled()
+  }
+  
+  func testParameterizedTupleClosure_anyWildcardMatching() {
+    given(concreteMock.parameterizedTupleClosure(block: any())) ~> true
+    XCTAssertTrue((concreteMock as ClosureParametersProtocol)
+      .parameterizedTupleClosure(block: { _ in }))
+    verify(concreteMock.parameterizedTupleClosure(block: any())).wasCalled()
+  }
+  
+  func testParameterizedReturningTupleClosure_anyWildcardMatching() {
+    given(concreteMock.parameterizedReturningTupleClosure(block: any())) ~> true
+    XCTAssertTrue((concreteMock as ClosureParametersProtocol)
+      .parameterizedReturningTupleClosure(block: { _ in fatalError() }))
+    verify(concreteMock.parameterizedReturningTupleClosure(block: any())).wasCalled()
+  }
+  
+  // MARK: Tuple wrapped escaping
+  
+  func testImplicitEscapingMultipleTupleClosure_anyWildcardMatching() {
+    given(concreteMock.implicitEscapingMultipleTupleClosure(block: any())) ~> true
+    XCTAssertTrue((concreteMock as ClosureParametersProtocol)
+      .implicitEscapingMultipleTupleClosure(block: ({}, { _ in fatalError() })))
+    verify(concreteMock.implicitEscapingMultipleTupleClosure(block: any())).wasCalled()
+  }
+  
+  func testImplicitEscapingMultipleLabeledTupleClosure_anyWildcardMatching() {
+    given(concreteMock.implicitEscapingMultipleLabeledTupleClosure(block: any())) ~> true
+    XCTAssertTrue((concreteMock as ClosureParametersProtocol)
+      .implicitEscapingMultipleLabeledTupleClosure(block: (a: {}, b: { _ in fatalError() })))
+    verify(concreteMock.implicitEscapingMultipleLabeledTupleClosure(block: any())).wasCalled()
+  }
+  
   
   // MARK: - notNil()
   

--- a/MockingbirdTestsHost/ClosureParameters.swift
+++ b/MockingbirdTestsHost/ClosureParameters.swift
@@ -49,4 +49,12 @@ protocol ClosureParametersProtocol {
   
   func escapingAutoclosureTrivialClosure(block: @escaping @autoclosure () -> Void) -> Bool
   func escapingAutoclosureTrivialReturningClosure(block: @escaping @autoclosure () -> Bool) -> Bool
+  
+  func trivialTupleClosure(block: (() -> Void)) -> Bool
+  func trivialReturningTupleClosure(block: (() -> Void)) -> Bool
+  func parameterizedTupleClosure(block: ((Bool) -> Void)) -> Bool
+  func parameterizedReturningTupleClosure(block: ((Bool) -> Bool)) -> Bool
+  
+  func implicitEscapingMultipleTupleClosure(block: (() -> Void, (Bool) -> Bool)) -> Bool
+  func implicitEscapingMultipleLabeledTupleClosure(block: (a: () -> Void, b: (Bool) -> Bool)) -> Bool
 }

